### ISSUE TITLE
Removing then changing iPhone PIN results in account passwords no longer working

### ISSUE THREAD

**Issue by rossbulat** [paritytech/parity-signer#1921]
I attempted to remove my iPhone's PIN code to speed up a demo I had planned for Decoded. 
After finding the app does not allow usage without a PIN, I set a 0000 PIN in the hope of again simplifying the demo flow. But what happened after this was that the app asks for a password when I attempt to scan a payload.
Both the old and new PIN did not work as the password, and so I had no choice but to re-install the app, and re-import the keysets.

**Comment by krodak** [comment]
Hi @rossbulat 
Described scenario is not an error, but expected device, let me explain.
We are using [kSecAttrAccessibleWhenPasscodeSetThisDeviceOnly ](https://developer.apple.com/documentation/security/ksecattraccessiblewhenpasscodesetthisdeviceonly) setting for Keychain API elements (here: seed phrases), due to security measures, to not be able to access those items when migrating to new device via backup, etc, which also means per Apple's documentation `No items can be stored in this class on devices without a passcode. Disabling the device passcode causes all items in this class to be deleted.`, so **device passcode is required to use Polkadot Vault**.

Case that you describe - not pincode not working for you - unfortunately this might be some issue on your end, as **Vault does not store any arbitrary pincode to access seed phrase data**, so pincode that you set up is **only maintained by your iOS device via Apple's system settings**. If newly set up pincode does not work for you when unlocking device / accessing keychain items (here: seed phrases), this means that you might have entered invalid pincode or tried to access items set up with previous pincode - which won't be available as per `kSecAttrAccessibleWhenPasscodeSetThisDeviceOnly` setting.

**Comment by rossbulat** [comment]
Thanks for explaining, that makes sense. I am pretty sure I attempted the previous PIN code and the new one, both of which didn't work. 

I hope you don't mind me re-opening so I can re-test this later & check whether both PIN codes do not work. If this is the case maybe it is worth amending the UI to remove the keysets & display a prompt to re-import them when PIN is changed. Or add a legacy group of keysets that are no longer accessible.

**Comment by krodak** [comment]
Sure, but please do mind that we don't have access to entered pin and we don't have overview of that operation as it happens outside of Vault and within iOS OS process - hence on our end, we do not validate PIN, we just get information from system callback whether your authentication attempt was successful or not 🙏🏻

**Comment by krodak** [comment]
>  If this is the case maybe it is worth amending the UI to remove the keysets & display a prompt to re-import them when PIN is changed.

Although it sounds tempting, iOS does not provide a public API for detecting when a user changes their device passcode (here: PIN). Apple highly values user privacy, and passcode changes are considered sensitive, user-specific information.

Therefore, we can't directly detect a passcode change. We can, however, indirectly infer that a passcode might have been changed by using the Keychain Services API. When we set the Keychain item accessibility to `kSecAttrAccessibleWhenPasscodeSetThisDeviceOnly`, the Keychain item becomes inaccessible and is removed if the user ever removes their device passcode - but this would mean that we'd need to try to access those items regularly, thus lowering security by accessing Keychain too often and triggering pincode prompt while doing so.

Moreover, this method is not foolproof. For example, it cannot distinguish between a pincode change where the pincode was temporarily removed and a situation where the user permanently removed their pincode. Finally, it can't detect the change of a passcode where the passcode was not removed in the process.
